### DISCUSSION TITLE
Ratio pomiędzy metrykami w Microsoft.Insights

### DISCUSSION THREAD

**Discussion by ad-m** [AzureCommunityPL/group#107] [Pomoc]
Hej,

Poznaje sobie `Microsoft.Insights`, jego możliwości, a także ograniczenia. Doceniam prostotę użycia względem alternatywnych self-hosted rozwiązań. 

Chciałbym sformułować alert, gdzie obliczam stosunek dwóch metryk, aby widzieć rezultat. Przykładowo liczba pomyślnych transakcji do ogólnej liczby transakcji (współczynnik awaryjności). Ewentualnie stosunek metryki `Ingress` do `Egress` w storageAccount.

Grafika, aby zwizualizować o czym mówię:
![obraz](https://user-images.githubusercontent.com/3618479/95582401-e6b5ff00-0a3a-11eb-9718-ba56d7ac180f.png)

Na dole widać:
- dwie metryki oznaczone jako `m1` i `m2` (odpowiednio `NumberOfNotificationsDelivered` i `NumberOfMessagesPublished`)
- metrykę "wirtualną" oznaczoną jako `e1`, która została wyliczona na podstawie wartości na podstawie `m1` i `m2`

Chciałbym stworzyć alert na odpowiedniku `e1`, ale obecnie widzę możliwość wyłącznie odwołanie do odpowiednika `m1` i `m2` w Azure. Potrzebuje pokierowania jak wyliczać wartości takie jak `e1`, aby stworzyć alert.

Podobne rozwiązanie zrealizowane z wykorzystaniem Zabbix:

![obraz](https://user-images.githubusercontent.com/3618479/95582896-9a1ef380-0a3b-11eb-9eee-66940e3a06fc.png)

Porównywany jest stosunek liczby możliwych do otwarcia plików `kernel.maxfiles.last` do liczby otwartych plików `vfs.file.regexp[/proc/sys/fs/file-nr,"([0-9]+)\s+([0-9]+)\s+([0-9]+)",,1,2,\1]`.

Jak należy to przełożyć na Azure, jeżeli chciałbym się zmigrować?
Z poważaniem,

**Comment by wi5nia** [comment]
Jesli chcesz to robic juz na poziomie apki to najlepiec bedzie [custom metrics](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/metrics-custom-overview) uzyc, a jak chcesz wyliczyc to pozniej z dwoch metryk, ktore wyslales to [zapytanie](https://docs.microsoft.com/en-us/azure/azure-monitor/log-query/get-started-queries) w Azure Monitorze

**Reply by ad-m** [reply]
Custom metrics wymaga zmiany po stronie aplikacji, a wolałbym to zrobić później, a w przypadku usług Azure – z natury usług SaaS  – nie kontroluje aplikacji, aby dodać własne metryki.
Czy "log queries", które podlinkowałeś nie bazuje na logach? Logi należy rozpatrywać jako podzbiór metryk? Wolałbym nie analizować logów, skoro mam już odpowiednie dane przeanalizowane w metrykach. Poza tym pisanie do logów np. ilość dostępnego miejsca na dysku...

**Reply by wi5nia** [reply]
To nie rozumiem jak chcesz się przełożyć na Azure nie korzystając z jego mechanizmów :) Azure Monitor obsługuje logi, [metryki](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-platform-metrics) - mój link wcześniej to był przykład do samego języka :)

Nie do końca chyba rozumiem Twoj problem i potrzebę w takim razie 🤷🏻‍♂️

**Reply by ad-m** [reply]
Ale ja chce korzystać z jego mechanizmów, ale w sposób możliwie efektywny. :) 

Proponujesz custom metrics. Czy custom metrics jestem w stanie wykorzystać np. dla Azure Storage? Czy w ten sposób jestem w stanie stworzyć np. trigger wyzwalany w zależności od ratio Ingress / Egress w Azure Storage?  

Proponujesz Azure Monitor. Ja wiem, że Azure Monitor obsługuje metryki i trochę znam usługę. Jednak metryki, a logi to – z tego co widzę – w tej usłudze dość mocno oddzielne rzeczy. Dla mnie idealnie byłoby zbudować "custom" metrykę na podstawie innych metryk, a nie metrykę na podstawie logów, albo samodzielnie stworzonej aplikacji (w Azure Storage "nie mam" aplikacji).

Przyjmijmy, że chce mieć alert, gdy ratio Ingress / Egress w Azure Storage przekracza określony stosunek (każda aplikacja ma swoją specyfikę ruchu, a anomalię powinny wzbudzać uwagę). Nie wiem jak to zrobić w "Azure-way" w prosty sposób. 

W AWS stworzyłbym "wirtualną" metrykę (podobną do widoku w SQL), która obliczałaby stosunek metryki "Ingress" do metryki "Egress", a następnie pozwoliłaby mi stworzyć odpowiedni wyzwalacz. Taka "wirtualna metryka" istnieje tylko odczytu danych i podczas definicji wyzwalacza i nie jest zapisywana jako "custom metrics" w AWS. 

W Zabbix w ramach definicji wyzwalacza określam wyrażenie odwołujące się do wielu metryk. Czyli byłoby to coś jak `{storage-123:Ingress} / {storage-123:Egress} > 0.5`.

Jak taki wyzwalacz robi się w Azure-way dla Azure Storage? Należy sięgać po logi? Należy samemu scrapować metryki z Azure Monitor, a następnie wysyłać je samemu z powrotem do Azure Monitor jako "custom metrics"?

**Comment by kamil-mrzyglod** [comment]
Nie testowałem nigdy tego, ale przynajmniej teoretycznie powinna być możliwość zrobienia tego czego szukasz. Generalnie potrzebujesz niestandardowej metryki opartej o standardowe. To o czym piszę @wi5nia zdałoby egzamin w przypadku appki, ale tutaj chyba trzeba nieco inaczej podejść do tematu. Alert o którym mówisz powinno dać radę stworzyć o ile wygenerujesz odpowiedni parametr. Można to zrobić pisząc własne zapytanie w Kusto w zakładce **Logs** w Azure Monitor i klikając na **+ New Alert rule**. To, czego nie będziesz prawdopodobnie mieć, to dane ze Storage Account. Tutaj z pomocą może przyjść eksport logów ze Storage Account do Log Analytics - https://microsoft.github.io/AzureTipsAndTricks/blog/tip205.html.

Mając to wszystko mógłbyś spróbować podejść do tego w sposób w który to opisujesz. Natywnego sposóbu na coś takiego niestety nie znam bo nie kojarzę możliwość w warunkach alertu z metrykami podania stosunku dwóch metryk 😉

**Reply by wi5nia** [reply]
Dokladnie, trzeba przez Kusto sobie albo bezposrednio w portalu zrobic query np w Workbooku albo odlozyc logi/metryki i je odpytac i na to ustawic alert.

**Reply by ad-m** [reply]
Dziękuje za informacje :) Teraz lepiej rozumiem jak modelować metryki, aby skutecznie używać Azure Monitor.

Brak wyrażeń w wyzwalaczach dużo zmienia w sposobie modelowania metryk, bo trzeba od razu myśleć jakie alerty chce się mieć nie zakładając, że wyzwalacz coś przeliczy, zamiast przedstawiania atomowych informacji.
